tuf/api/metadata.py

```diff
@@ -1258,17 +1267,173 @@ def to_dict(self) -> Dict[str, Any]:
+            **self.unrecognized_fields,
+        }
+
+    def get_bin_name(self, bin_number: int = 0) -> str:
+        """Return the name of bin with number ``bin_number``.
+        The ``bin_number`` default is 0 as that is used to uniquely identify
+        delegated roles containing succinct hash delegations from the rest.
+
+        Args:
+            bin_number: Number of the bin that we want the name for.
+        """
+        # Add zero padding if necessary.
+        suffix = f"{bin_number:0{self.zero_padding}}"
+        return f"{self.bin_name_prefix}-{suffix}"
+
+    def get_all_bin_names(self) -> Iterator[str]:
+        """Yield the names of all different bins one by one."""
+        for i in range(0, self.number_of_bins):
+            yield self.get_bin_name(i)
+
+    def is_bin(self, role_name: str) -> bool:
+        """Determines whether the given ``role_name`` is in one of
+        the bins that ``SuccinctHashDelegations`` is trusted to provide.
+
+        Args:
+            role_name: The name of the delegated role to check against.
+        """
+        desired_prefix = self.bin_name_prefix + "-"
+
+        if role_name.startswith(desired_prefix):
+            suffix = role_name[len(desired_prefix) :]
+            try:
+                num = int(suffix)
+                if 0 <= num < self.number_of_bins:
+                    return True
+            except ValueError:
+                # suffix is not a number
+                return False
+
+        return False
+
+    def _find_bin_for_bits(self, hash_bits_representation: str) -> str:
+        """Helper function for find_delegation calculating the actual rolename.
+
+        Args:
+            hash_bits_representation: binary bit representation of the target
+                hash.
+        """
+
+        bit_length = self.hash_prefix_len
+        # Get the first bit_length of bits and then cast them to decimal.
+        bin_number = int(hash_bits_representation[:bit_length], 2)
+
+        return self.get_bin_name(bin_number)
+
+    def find_bin(self, target_filepath: str) -> str:
+        """Calculates the name of the bin responsible for ``target_filepath``.
+
+        Args:
+            target_filepath: URL path to a target file, relative to a base
+                targets URL.
+        """
+        hasher = sslib_hash.digest(algorithm="sha256")
+        hasher.update(target_filepath.encode("utf-8"))
+        # Get the bit representation of the hash. The bit representation is
+        # calculated for a small number of symbols required by hash_prefix_len.
+        needed_bits = self.hash_prefix_len  # aliased for emphasis
+        needed_bytes = math.ceil(needed_bits / 8)
+        hash_bits = "".join(
+            f"{one_byte:08b}" for one_byte in hasher.digest()[:needed_bytes]
+        )
+        needed_hash_bits = hash_bits[:needed_bits]
+
+        # Split hash calculation from the rest for easier testing.
+        return self._find_bin_for_bits(needed_hash_bits)
```

Comment:
I've not played with this yet but after looking at this and talking to Lukas, I think I finally understood what's being done here: it's essentially a right shift to get hash_prefix_len leftmost bits. I think if this is the case then I would prefer seeing the actual math here.

This is entirely untested but: 
```
h = hasher.digest()[:8] # we can't ever need more than 8 digits
# right shift so that we only have the bits we care about
bin_number = int(h, 16) >> (8*4 - self.hash_prefix_len)
return self.get_bin_name(bin_number)
```

Reply:
🤦  "Obviously"  bit shifting is the way to do this.

Reply:
You are correct that directly shifting the bits is easier to understand.
Will apply this. Thanks for the detailed comment!